src/mgr/QueryParams.cc

```diff
@@ -65,61 +69,69 @@ Mgr::QueryParams::find(const String& name) const
     return iter;
 }
 
-bool
-Mgr::QueryParams::ParseParam(const String& paramStr, Param& param)
+/// Parses the value part of a "param=value" URL section.
+/// Value can be a comma-separated list of integers or an opaque string.
+/// \note opaque string may be a list with a non-integer (e.g., "1,2,3,z")
+Mgr::QueryParam::Pointer
+ParseParamValue(SBuf &rawValue)
 {
-    bool parsed = false;
-    regmatch_t pmatch[3];
-    regex_t intExpr;
-    regcomp(&intExpr, "^([a-z][a-z0-9_]*)=([0-9]+((,[0-9]+))*)$", REG_EXTENDED | REG_ICASE);
-    regex_t stringExpr;
-    regcomp(&stringExpr, "^([a-z][a-z0-9_]*)=([^&= ]+)$", REG_EXTENDED | REG_ICASE);
-    if (regexec(&intExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        std::vector<int> array;
-        int n = pmatch[2].rm_so;
-        for (int i = n; i < pmatch[2].rm_eo; ++i) {
-            if (paramStr[i] == ',') {
-                array.push_back(atoi(paramStr.substr(n, i).termedBuf()));
-                n = i + 1;
-            }
-        }
-        if (n < pmatch[2].rm_eo)
-            array.push_back(atoi(paramStr.substr(n, pmatch[2].rm_eo).termedBuf()));
-        param.second = new IntParam(array);
-        parsed = true;
-    } else if (regexec(&stringExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        param.second = new StringParam(paramStr.substr(pmatch[2].rm_so, pmatch[2].rm_eo));
-        parsed = true;
+    static const CharacterSet comma("comma", ",");
+
+    Parser::Tokenizer tok(rawValue);
+    std::vector<int> array;
+    int64_t intVal = 0;
+    while (tok.int64(intVal, 10, false)) {
+        Must(intVal >= std::numeric_limits<int>::min());
+        Must(intVal <= std::numeric_limits<int>::max());
+        array.emplace_back(intVal);
+        if (tok.remaining().length() > 1)
```

Comment:
Why do we only want to consume a comma if there are at least two characters left. If this strange condition is not needed, let's remove it. If it is needed, then please add a source code comment to explain what is going on.

```suggestion
```

Reply:
Now that we are forbidding empty list entries this is needed to prevent the list ending on a comma.

Reply:
IMHO, it is better to allow a single trailing comma, like many modern languages do, than to complicate code and add source code comments to explain this condition. As an added bonus, allowing a trailing comma will simplify generation of cache manager URLs. Your call though. I see three valid options here:

1. Allow a single trailing comma: Consume the comma unconditionally, perhaps with a trailing `// allow a trailing comma` comment.
2. Ban trailing commas, with an explanation: Keep this condition. Add a source code comment to explain what is going on here.
3. Ban trailing commas, without an explanation: Refactor the code to make it self-documenting. Remove the condition and save skipOne(comma) result in a new sawCommaAtTheEnd boolean variable, to be used together with the atEnd() condition below.

Reply:
Im going with not changing the API. 
a26458a documents the pattern being parsed. I hope that is clear enough now.

Reply:
> I hope that is clear enough now.

Unfortunately, it is not clear enough to me. The added comment does not explain the strange condition. In my experience, it is conditions like this that sooner or later break "working" parsing code or make it difficult to maintain.

Reply:
How about the 4451407 texts?

Reply:
The primary problem -- strange parsing conditions often lead to bugs -- remains, of course. The secondary problem -- it is difficult to understand what the code is trying to do -- can be considered addressed, albeit in a rather tortured/convoluted (IMO) and sloppy way. I do not have enough time to properly address these non-blocking problems myself, so I will resolve this change request.